pkg/controller/common/helm/patch.go

```diff
@@ -35,6 +37,17 @@ type basicPatch struct {
 }
 
 func (p *basicPatch) Apply(ctx context.Context) (*unstructured.Unstructured, error) {
+	if p.oldObj.GroupVersionKind().Group == "route.openshift.io" && p.oldObj.GroupVersionKind().Kind == "Route" &&
+		routeHostChanged(p.oldObj, p.newObj) {
+		return nil, &errors.StatusError{
+			ErrStatus: metav1.Status{
+				Status:  metav1.StatusFailure,
+				Reason:  metav1.StatusReasonInvalid,
+				Message: "spec.host of Route objects is immutable",
+			},
+		}
+	}
+
```

Comment:
in fact, the `Patch()` would always fail anyway (server-side) when changing `spec.host` (because it's immutable), triggering a recreate. this just lets it fail earlier.

Reply:
Are you sure it's immutable? I tried changing an existing route's host with `kubectl edit` after reading the error message above, and it worked.

Reply:
Okay, that's strange. I have to test it again, but it's documented as "may not be changed": https://github.com/openshift/api/blob/803c45de7ab5567e8d1575138f014226974768a1/route/v1/types.go#L21-L22

Reply:
Oh you are right. You can change it. Weird, I was somehow sure that `spec.host` can't be changed.

Another thing I noticed is that the auto-generated `host` value is actually written to the `spec` (likely using a MutatingWebhook), and an annotation is added: `openshift.io/host.generated: "true"`. So we might have to check that annotation and remove it

Reply:
I don't think we need to update the annotation. OpenShift should do this (there was a [TODO](https://github.com/openshift/origin/blob/95117a2f8866f369b1ac2828ee520975c09593bc/pkg/route/registry/route/rest.go#L136) about this, but I guess they never actually did it.

AFAICT that annotation is only used to show the route host as being auto-generated in the OpenShift console, so it's not a big deal if we leave it as is.